internal/juju/applications.go

```diff
@@ -996,31 +996,90 @@ func (c applicationsClient) ReadApplication(input *ReadApplicationInput) (*ReadA
 
 	appInfo := apps[0].Result
 
-	// Fetch data only about the application being read. This helps to limit
-	// the data on storage to the specific application too. Storage is not
-	// provided by application, rather storage data buries a unit name deep
+	// Fetch status. Storage is not provided by application,
+	// rather storage data buries a unit name deep
 	// in the structure.
+	// TODO(alesstimec): Switch to using GetApplicationStorage once juju 4 implements it.
 	status, err := clientAPIClient.Status(&apiclient.StatusArgs{
-		Patterns:       []string{input.AppName},
 		IncludeStorage: true,
 	})
 	if err != nil {
-		if strings.Contains(err.Error(), "filesystem for storage instance") ||
-			strings.Contains(err.Error(), "volume for storage instance") ||
-			strings.Contains(err.Error(), "cannot convert storage details") {
-			// Retry if we get this error. It means the storage is not ready yet.
-			return nil, NewStorageNotFoundError(input.AppName)
-		}
 		c.Errorf(err, "failed to get status")
 		return nil, err
 	}
+
 	var appStatus params.ApplicationStatus
 	var exists bool
 	if appStatus, exists = status.Applications[input.AppName]; !exists {
 		return nil, fmt.Errorf("no status returned for application: %s", input.AppName)
 	}
 
-	storages := c.transformToStorageConstraints(status.Storage, status.Filesystems, status.Volumes)
+	// because we cannot rely on the Juju controller to filter storage/volumes/filesystems for us (looking at Juju 4)
+	// we need to filter here.
+
+	// first we collect all application units and machines
+	appUnits := make(map[string]bool)
+	appMachines := make(map[string]bool)
+	for unitTag, unitStatus := range appStatus.Units {
+		appUnits[unitTag] = true
+		appMachines[unitStatus.Machine] = true
+	}
+
+	// then do the filtering
+	appStorage := []params.StorageDetails{}
+	appFilesystems := []params.FilesystemDetails{}
+	appVolumes := []params.VolumeDetails{}
+	for _, storageDetails := range status.Storage {
+		isAppStorage := false
+		for unitTag, _ := range storageDetails.Attachments {
+			if appUnits[unitTag] {
+				isAppStorage = true
+				break
+			}
+		}
+		if isAppStorage {
+			appStorage = append(appStorage, storageDetails)
+		}
+	}
+
+	for _, filesystemDetails := range status.Filesystems {
+		isAppFilesystem := false
+		for machineTag, _ := range filesystemDetails.MachineAttachments {
+			if appMachines[machineTag] {
+				isAppFilesystem = true
+				break
+			}
+		}
+		for unitTag, _ := range filesystemDetails.UnitAttachments {
+			if appUnits[unitTag] {
+				isAppFilesystem = true
+				break
+			}
+		}
+		if isAppFilesystem {
+			appFilesystems = append(appFilesystems, filesystemDetails)
+		}
+	}
+
+	for _, volumeDetails := range status.Volumes {
+		isAppVolume := false
+		for machineTag, _ := range volumeDetails.MachineAttachments {
+			if appMachines[machineTag] {
+				isAppVolume = true
+				break
+			}
+		}
+		for unitTag, _ := range volumeDetails.UnitAttachments {
```

Comment:
Don't we just care about the unit here? Could we end up with duplicates if the same volume is on machine A for unit A?